libgnucash/app-utils/gnc-option.hpp

```diff
@@ -0,0 +1,254 @@
+/********************************************************************\
+ * gnc-option.hpp -- Application options system                     *
+ * Copyright (C) 2020 John Ralls <jralls@ceridwen.us>               *
+ *                                                                  *
+ * This program is free software; you can redistribute it and/or    *
+ * modify it under the terms of the GNU General Public License as   *
+ * published by the Free Software Foundation; either version 2 of   *
+ * the License, or (at your option) any later version.              *
+ *                                                                  *
+ * This program is distributed in the hope that it will be useful,  *
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of   *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the    *
+ * GNU General Public License for more details.                     *
+ *                                                                  *
+ * You should have received a copy of the GNU General Public License*
+ * along with this program; if not, contact:                        *
+ *                                                                  *
+ * Free Software Foundation           Voice:  +1-617-542-5942       *
+ * 51 Franklin Street, Fifth Floor    Fax:    +1-617-542-2652       *
+ * Boston, MA  02110-1301,  USA       gnu@gnu.org                   *
+ *                                                                  *
+\********************************************************************/
+/** @addtogroup Engine
+    @{ */
+/** @addtogroup Options
+    @{ */
+
+/** @file gnc-option.hpp
+    @brief C++ Public interface for individual options.
+    @author Copyright 2020-2021 John Ralls <jralls@ceridwen.us>
+*/
+
+#ifndef GNC_OPTION_HPP_
+#define GNC_OPTION_HPP_
+
+#include <glib.h>
+#include <libguile.h>
+#include <string>
+#include <iostream>
+#include <iomanip>
+#include <variant>
+#include <memory>
+#include "gnc-option-ui.hpp"
+#include "gnc-option-date.hpp"
+
+struct OptionClassifier;
+class GncOptionUIItem;
+using GncOptionUIItemPtr = std::unique_ptr<GncOptionUIItem>;
+#ifndef SWIG //SWIG pulls in GncOwner from swig-engine.
+struct _gncOwner;
+using GncOwner = _gncOwner;
+#endif
+struct _QofQuery;
+using QofQuery = _QofQuery;
+struct QofInstance_s;
+using QofInstance = QofInstance_s;
+template <typename ValueType> class GncOptionValue;
+class GncOptionAccountListValue;
+class GncOptionAccountSelValue;
+class GncOptionMultichoiceValue;
+template <typename ValueType> class GncOptionRangeValue;
+template <typename ValueType> class GncOptionValidatedValue;
+class GncOptionDateValue;
+
+template <typename T>
+struct is_OptionClassifier
+{
+    static constexpr bool value =
+        std::is_base_of_v<OptionClassifier, std::decay_t<T>>;
+};
+
+template <typename T> inline constexpr bool
+is_OptionClassifier_v = is_OptionClassifier<T>::value;
```

Comment:
Interesting constructs and I get more or less what these are doing. I don't see though why this is done in two steps - first defining a struct that sets a value and then an assignment to a _v named variable.
Do you have pointers to documentation on this I can study ?

Reply:
Well, the documentation I used is [Josuttis & Vandevoord. C++ Templates: The Complete Guide](http://www.tmplbook.com), but you might already have some other book with a discussion of SFINAE games. This particular two-step was introduced to the standard library in C++17. It's pure syntactic sugar and I duplicate it for these summary template classes so that the uses are consistent.

Reply:
Ok. I had heard the term SFINAE but never actually studied it. It's new to me you can have structs that only define types rather than having member variables. A whole new world to consider...